### PR TITLE
Add part: lstm block

### PR DESCRIPTION
**Pull request by christophmluscher**
Adds LSTM Block

### CODE REVIEW

File: i6_models/parts/lstm.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    input_dim: int
+    hidden_dim: int
+    num_layers: int
+    bias: bool
+    dropout: float
+    bidirectional: bool
+    enforce_sorted: bool
+
+    @classmethod
+    def from_dict(cls, model_cfg_dict: Dict):
```
Comment (NeoLegends) [inline i6_models/parts/lstm.py:22]: Same Q as in the other PR: why is this necessary now, and hasn't been for the other assemblies?

**Review by NeoLegends — APPROVED** [review]
One minor Q.

File: i6_models/parts/lstm.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            input_size=self.cfg.input_dim,
+            hidden_size=self.cfg.hidden_dim,
+            num_layers=self.cfg.num_layers,
+            bias=self.cfg.bias,
+            dropout=self.dropout,
+            batch_first=True,
+            bidirectional=self.cfg.bidirectional,
+        )
+
+    def forward(self, x: torch.Tensor, seq_len: torch.Tensor) -> (torch.Tensor, torch.Tensor):
```
Comment (albertz) [inline i6_models/parts/lstm.py:45]:
```suggestion
    def forward(self, x: torch.Tensor, seq_len: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```

File: i6_models/parts/lstm.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    enforce_sorted: bool
+
+    @classmethod
+    def from_dict(cls, model_cfg_dict: Dict):
+        model_cfg_dict = model_cfg_dict.copy()
+        return cls(**model_cfg_dict)
+
+
+class LstmBlockV1(nn.Module):
+    def __init__(self, model_cfg: Union[LstmBlockV1Config, Dict], **kwargs):
```
Comment (albertz) [inline i6_models/parts/lstm.py:28]:
```suggestion
    def __init__(self, model_cfg: Union[LstmBlockV1Config, Dict[str, Any]], **kwargs):
```

File: i6_models/parts/lstm.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            bias=self.cfg.bias,
+            dropout=self.dropout,
+            batch_first=True,
+            bidirectional=self.cfg.bidirectional,
+        )
+
+    def forward(self, x: torch.Tensor, seq_len: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        if not torch.jit.is_scripting() and not torch.jit.is_tracing():
+            if seq_len.get_device() >= 0:
+                seq_len = seq_len.cpu()
```
Comment (albertz) [inline i6_models/parts/lstm.py:54]:
```suggestion
            seq_len = seq_len.cpu()
```

File: i6_models/parts/lstm.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            hidden_size=self.cfg.hidden_dim,
+            num_layers=self.cfg.num_layers,
+            bias=self.cfg.bias,
+            dropout=self.dropout,
+            batch_first=True,
+            bidirectional=self.cfg.bidirectional,
+        )
+
+    def forward(self, x: torch.Tensor, seq_len: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        if not torch.jit.is_scripting() and not torch.jit.is_tracing():
```
Comment (albertz) [inline i6_models/parts/lstm.py:52]: Why only when not scripting? Don't you want that `seq_len` is always on CPU?

Comment (christophmluscher) [reply i6_models/parts/lstm.py:52]: I followed the example in the blstm part.
```
        if not torch.jit.is_scripting() and not torch.jit.is_tracing():
            # during graph mode we have to assume all Tensors are on the correct device,
            # otherwise move lengths to the CPU if they are on GPU
            if seq_len.get_device() >= 0:
                seq_len = seq_len.cpu()
```
I did not copy the comment over... I did not yet get to look why this is necessary
@JackTemaki you implemented the BLSTM IIRC. You remember why this was done in this way?

File: i6_models/parts/lstm.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+from dataclasses import dataclass
+import torch
+from torch import nn
+from typing import Dict, Tuple, Union
+
+from i6_models.config import ModelConfiguration
+
+
+@dataclass
+class LstmBlockV1Config(ModelConfiguration):
```
Comment (Atticus1806) [inline i6_models/parts/lstm.py:12]: Could you add docs here please? You can probably copy them from Blst

File: i6_models/parts/lstm.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    input_dim: int
+    hidden_dim: int
+    num_layers: int
+    bias: bool
+    dropout: float
+    bidirectional: bool
+    enforce_sorted: bool
+
+    @classmethod
+    def from_dict(cls, model_cfg_dict: Dict[str, Any]):
```
Comment (Atticus1806) [inline i6_models/parts/lstm.py:22]: I don't see this for other part configs, why do we need this here?

File: i6_models/parts/lstm.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+
+@dataclass
+class LstmBlockV1Config(ModelConfiguration):
+    input_dim: int
+    hidden_dim: int
+    num_layers: int
+    bias: bool
+    dropout: float
+    bidirectional: bool
```
Comment (Atticus1806) [inline i6_models/parts/lstm.py:18]: should we allow this? I feel like if we have bidirectional here, the BLSTM part becomes redundant, which is maybe okay, but might also cause two different branches that do the same, which I am not sure we want (if there are potential extensions later). We could maybe also just deprecate the BLSTM block?

File: i6_models/parts/lstm.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    @classmethod
+    def from_dict(cls, model_cfg_dict: Dict[str, Any]):
+        model_cfg_dict = model_cfg_dict.copy()
+        return cls(**model_cfg_dict)
+
+
+class LstmBlockV1(nn.Module):
+    def __init__(self, model_cfg: Union[LstmBlockV1Config, Dict[str, Any]], **kwargs):
+        """
+        Model definition of LSTM block. Contains single lstm stack and padding sequence in forward call.
```
Comment (Atticus1806) [inline i6_models/parts/lstm.py:30]: Also add the "including dropout, batch-first variant,  hardcoded to use B,T,F input" part please.

Comment (Atticus1806) [reply i6_models/parts/lstm.py:30]: Could also add the supports scripting part.

File: i6_models/parts/lstm.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+            input_size=self.cfg.input_dim,
+            hidden_size=self.cfg.hidden_dim,
+            num_layers=self.cfg.num_layers,
+            bias=self.cfg.bias,
+            dropout=self.dropout,
+            batch_first=True,
+            bidirectional=self.cfg.bidirectional,
+        )
+
+    def forward(self, x: torch.Tensor, seq_len: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (Atticus1806) [inline i6_models/parts/lstm.py:51]: doc pls.

File: i6_models/parts/lstm.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+        lstm_packed_in = nn.utils.rnn.pack_padded_sequence(
+            input=x,
+            lengths=seq_len,
+            enforce_sorted=self.enforce_sorted,
+            batch_first=True,
+        )
+
+        lstm_out, _ = self.lstm_stack(lstm_packed_in)
+        lstm_out, _ = nn.utils.rnn.pad_packed_sequence(
```
Comment (Atticus1806) [inline i6_models/parts/lstm.py:64]: Just out of curiosity: why does black force the new lines here but not for the blstm? Shouldnt it be the same line length?

**Review by Atticus1806 — CHANGES_REQUESTED** [review]

File: i6_models/parts/lstm.py (hunk 5 above)
Comment (JackTemaki) [reply i6_models/parts/lstm.py:52]: The question is, is this still relevant? This was something I added at some point, but if this is not needed for ONNX export this should be removed until there is actually a reason for it.

File: i6_models/parts/lstm.py (hunk 7 above)
Comment (christophmluscher) [reply i6_models/parts/lstm.py:22]: I use it in the model definition to conert a dict to the config class. Need might be a bit strong but I like it :D

File: i6_models/parts/lstm.py (hunk 11 above)
Comment (christophmluscher) [reply i6_models/parts/lstm.py:64]: well this is dependent if you manually set the last commata. if set it will force new lines

File: i6_models/parts/lstm.py (hunk 1 above)
Comment (christophmluscher) [reply i6_models/parts/lstm.py]: I use it in the model definition to convert a dict to the config class. Necessary might be a bit strong but I like it :D

**Review by Atticus1806 — APPROVED** [review]
